### PR TITLE
[Indexer] Client Implementation & Bug Fixes

### PR DESCRIPTION
**Pull request by ethenotethan**
**Description**
- Fixed an API bug where deposits and withdrawals were being returned in response payloads for unspecified addresses. The two deposit/withdrawal SQL queries take a union of two SQL select statements. The second statement had no `from_address` filter which resulted in unintended rows being returned. 
- Fixed error response bug where failed requests where met with an unformatted error message on-top of a blank payload. 
- Added validation abstraction to API that performs request param integrity checks for address, cursor, & limit values. 
- Added extensive code documentation to existing implementation.
- Added a lightweight API client package for programmatically querying the API via go. 
- Updated API object to take in a listener that exposes a binded port to other modules. This is used for ensuring unique port allocation for each E2E test. 
- Moved data models into `api/models` package for better abstraction decoupling
- Updated typescript client bindings

**Tests**
- Tested validator functions
- Added E2E test that ensures that withdrawals and deposits can be extracted via API client
- Extended E2E test suite to take in API construct and client 

**Additional context**

Add any other context about the problem you're solving.

**Metadata**

- Fixes #[Link to Issue]

### CODE REVIEW

File: indexer/api/api.go [hunk 1]
```diff
@@ -20,45 +20,61 @@ import (
...
 }
 
 const (
 	MetricsNamespace = "op_indexer"
+	addressParam     = "{address:%s}"
+
+	// Endpoint paths
+	HealthPath      = "/healthz"
+	DepositsPath    = "/api/v0/deposits/"
+	WithdrawalsPath = "/api/v0/withdrawals/"
```
Comment (roninjin10) [inline indexer/api/api.go:42]: I like pulling this out to constants. We could even consider pulling the version out into it's own constant

File: indexer/api/routes/validator.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+	"errors"
+)
+
+// Validator ... Validates API user request parameters
+type Validator struct {
+}
+
+// ValidateQueryParams ... Validates the limit and cursor query parameters
+func (v *Validator) ValidateLimit(limit string) (int, error) {
```
Comment (roninjin10) [inline indexer/api/routes/validator.go]: nice!

File: indexer/client/client.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+const (
+	urlParams = "?cursor=%s&limit=%d"
+
+	defaultPagingLimit = 100
+)
+
+// Config ... Indexer client config struct
+type Config struct {
+	PaginationLimit int
+	URL             string
```
Comment (roninjin10) [inline indexer/client/client.go]: nit I think more clear if this is named baseurl

File: indexer/e2e_tests/bridge_transfers_e2e_test.go [hunk 4]
```diff
@@ -445,3 +447,76 @@ func TestE2EBridgeTransfersCursoredWithdrawals(t *testing.T) {
 		require.Equal(t, int64(3-i)*params.Ether, withdrawal.L2BridgeWithdrawal.Tx.Amount.Int64())
 	}
 }
+
+func Test_ClientGetWithdrawals(t *testing.T) {
```
Comment (roninjin10) [inline indexer/e2e_tests/bridge_transfers_e2e_test.go]: nice!

**Review by roninjin10 — COMMENTED** [review]
This is awesome! I just skimmed it since it's still a draft

File: indexer/client/client.go (hunk 3 above)
Comment (ethenotethan) [reply indexer/client/client.go]: Agreed: [75c6231](https://github.com/ethereum-optimism/optimism/pull/7312/commits/75c623151365909c672ed4e1624257caa912eb19)

File: indexer/api/routes/deposits.go [hunk 5]
```diff
@@ -46,39 +28,55 @@ func newDepositResponse(deposits *database.L1BridgeDepositsResponse) DepositResp
...
+		h.logger.Error("Invalid address param", "param", addressValue)
+		h.logger.Error(err.Error())
+		return
+	}
+
+	err = h.v.ValidateCursor(cursor)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		h.logger.Error("Invalid cursor param", "param", cursor)
+		h.logger.Error(err.Error())
```
Comment (hamdiallam) [inline indexer/api/routes/deposits.go]:
```suggestion
		h.logger.Error("Invalid cursor param", "param", cursor, "err", error.Error())
```

File: indexer/api/routes/deposits.go [hunk 6]
```diff
@@ -46,39 +28,55 @@ func newDepositResponse(deposits *database.L1BridgeDepositsResponse) DepositResp
...
+		return
 	}
 
-	deposits, err := h.BridgeTransfersView.L1BridgeDepositsByAddress(address, cursor, limit)
+	limit, err := h.v.ParseValidateLimit(limitQuery)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		h.logger.Error("Invalid limit param", "param", limitQuery)
+		h.logger.Error(err.Error())
```
Comment (hamdiallam) [inline indexer/api/routes/deposits.go]:
```suggestion
		h.logger.Error("Invalid limit param", "param", limitQuery, "err", err.Error())
```

File: indexer/api/routes/deposits.go [hunk 7]
```diff
@@ -46,39 +28,55 @@ func newDepositResponse(deposits *database.L1BridgeDepositsResponse) DepositResp
...
+		return
+	}
+
+	deposits, err := h.view.L1BridgeDepositsByAddress(address, cursor, limit)
 	if err != nil {
 		http.Error(w, "Internal server error reading deposits", http.StatusInternalServerError)
-		h.Logger.Error("Unable to read deposits from DB")
-		h.Logger.Error(err.Error())
+		h.logger.Error("Unable to read deposits from DB")
+		h.logger.Error(err.Error())
```
Comment (hamdiallam) [inline indexer/api/routes/deposits.go]:
```suggestion
		h.logger.Error("Unable to read deposits from DB", "err", err.Error())
```

File: indexer/api/routes/validator_test.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
+package routes
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (hamdiallam) [inline indexer/api/routes/validator_test.go:8]: Small nit just for consistent: Snake case for test names `TestParseValidateLimit`

File: indexer/api/routes/withdrawals.go [hunk 9]
```diff
@@ -48,36 +28,53 @@ func newWithdrawalResponse(withdrawals *database.L2BridgeWithdrawalsResponse) Wi
...
-			http.Error(w, "Limit could not be parsed into a number", http.StatusBadRequest)
-			h.Logger.Error("Invalid limit")
-			h.Logger.Error(err.Error())
-		}
-		limit = parsedLimit
+	address, err := h.v.ParseValidateAddress(addressValue)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		h.logger.Error("Invalid address param", "param", addressValue)
+		h.logger.Error(err.Error())
```
Comment (hamdiallam) [inline indexer/api/routes/withdrawals.go]:
```suggestion
		h.logger.Error("Invalid address param", "param", addressValue, "err", err)
```

File: indexer/e2e_tests/bridge_transfers_e2e_test.go (hunk 4 above)
Comment (hamdiallam) [reply indexer/e2e_tests/bridge_transfers_e2e_test.go]: Bump here about snake case with the test name

File: indexer/e2e_tests/setup.go [hunk 10]
```diff
@@ -105,14 +118,54 @@ func createE2ETestSuite(t *testing.T) E2ETestSuite {
...
+		}
+	}()
+
+	t.Cleanup(func() {
+		apiStop()
+		indexerStop()
+	})
+
+	// Wait for the API to start listening
+	time.Sleep(1 * time.Second)
```
Comment (hamdiallam) [inline indexer/e2e_tests/setup.go:157]: ffti: can also wait for the health check endpoint to return 200

**Review by hamdiallam — APPROVED** [review]
Looks great! Just left a few comments

File: .vscode/extensions.json [hunk 11]
```diff
@@ -1,12 +1,12 @@
 {
-  // See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
-  // Extension identifier format: ${publisher}.${name}. Example: vscode.csharp
+    // See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
```
Comment (hamdiallam) [inline .vscode/extensions.json]: Prolly want to undo the .vscode changes

File: indexer/e2e_tests/setup.go (hunk 10 above)
Comment (ethenotethan) [reply indexer/e2e_tests/setup.go:157]: That is true but the port would be unknown for some duration of time which is why we're doing this wait here

File: .vscode/extensions.json (hunk 11 above)
Comment (ethenotethan) [reply .vscode/extensions.json]: 2a8dc03

File: indexer/api/routes/withdrawals.go (hunk 9 above)
Comment (ethenotethan) [reply indexer/api/routes/withdrawals.go]: 2a8dc03

File: indexer/api/routes/validator_test.go (hunk 8 above)
Comment (ethenotethan) [reply indexer/api/routes/validator_test.go:8]: 2a8dc03

File: indexer/api/routes/deposits.go (hunk 7 above)
Comment (ethenotethan) [reply indexer/api/routes/deposits.go]: 2a8dc03

File: indexer/api/routes/deposits.go (hunk 6 above)
Comment (ethenotethan) [reply indexer/api/routes/deposits.go]: 2a8dc03

File: indexer/api/routes/deposits.go (hunk 5 above)
Comment (ethenotethan) [reply indexer/api/routes/deposits.go]: 2a8dc03

**Pull request merged**